File: skills/motion-world/references/runtimes/android/FrameSequence.kt
package motionworld

import kotlin.math.roundToInt

object FrameSequence {
    fun index(progress: Float, frameCount: Int, reducedMotion: Boolean = false): Int {
        val count = frameCount.coerceAtLeast(1)
        var p = MotionProgress.clamp(progress)
        if (reducedMotion) p = if (p < 0.25f) 0f else if (p < 0.75f) 0.5f else 1f
        return (p * (count - 1)).roundToInt().coerceIn(0, count - 1)
    }

    fun fileName(progress: Float, frameCount: Int, reducedMotion: Boolean = false): String =
        "frame_%04d.webp".format(index(progress, frameCount, reducedMotion))
}

File: skills/motion-world/references/runtimes/flutter/frame_sequence.dart
import 'motion_progress.dart';

class FrameSequence {
  static int index(double progress, int frameCount, {bool reducedMotion = false}) {
    final count = frameCount < 1 ? 1 : frameCount;
    var p = MotionProgress.clamp(progress);
    if (reducedMotion) p = p < 0.25 ? 0 : (p < 0.75 ? 0.5 : 1);
    return (p * (count - 1)).round().clamp(0, count - 1);
  }

  static String fileName(double progress, int frameCount, {bool reducedMotion = false}) {
    final i = index(progress, frameCount, reducedMotion: reducedMotion);
    return 'frame_${i.toString().padLeft(4, '0')}.webp';
  }
}
